src/utils/PSMFeatureExtractor.cpp

```diff
@@ -0,0 +1,304 @@
+      }
+      else
+      {
+        String search_engine = protein_ids.front().getSearchEngine();
+        if (!ListUtils::contains(search_engines_used, search_engine))
+        {
+          search_engines_used.push_back(search_engine);
+        }
+        
+        if (concatenate)
+        {
+          // will concatenate the list
+          PercolatorFeatureSetHelper::concatMULTISEPeptideIds(all_peptide_ids, peptide_ids, search_engine);
+        }
+        else
+        {
+          // will collapse the list (reference)
+          PercolatorFeatureSetHelper::mergeMULTISEPeptideIds(all_peptide_ids, peptide_ids, search_engine);
+        }
+      }
+      PercolatorFeatureSetHelper::mergeMULTISEProteinIds(all_protein_ids, protein_ids);
+    }
+    
+    if (all_protein_ids.empty())
+    {
+      writeLog_("No protein hits found in input file. Aborting!");
+      printUsage_();
+      return INPUT_FILE_EMPTY;
+    }
+
+    //-------------------------------------------------------------
+    // extract search engine and prepare pin
+    //-------------------------------------------------------------
+    String search_engine = all_protein_ids.front().getSearchEngine();
+    if (multiple_search_engines) search_engine = "multiple";
+    LOG_DEBUG << "Registered search engine: " << search_engine << endl;
+    
+    StringList extra_features = getStringList_("extra");
+    StringList feature_set;
+
+    if (search_engine == "multiple")
+    {
+      if (getFlag_("concat"))
+      {
+        PercolatorFeatureSetHelper::addCONCATSEFeatures(all_peptide_ids, search_engines_used, feature_set);
+      }
+      else
+      {
+        bool impute = getFlag_("impute");
+        bool limits = getFlag_("limit_imputation");
+        PercolatorFeatureSetHelper::addMULTISEFeatures(all_peptide_ids, search_engines_used, feature_set, !impute, limits);
+      }
+    }
+    else if (search_engine == "MS-GF+") PercolatorFeatureSetHelper::addMSGFFeatures(all_peptide_ids, feature_set);
+    else if (search_engine == "Mascot") PercolatorFeatureSetHelper::addMASCOTFeatures(all_peptide_ids, feature_set);
+    else if (search_engine == "XTandem") PercolatorFeatureSetHelper::addXTANDEMFeatures(all_peptide_ids, feature_set);
+    else if (search_engine == "Comet") PercolatorFeatureSetHelper::addCOMETFeatures(all_peptide_ids, feature_set);
+    else
+    {
+      writeLog_("No known input to create PSM features from. Aborting");
+      return INCOMPATIBLE_INPUT_DATA;
+    }
+
+    String run_identifier = all_protein_ids.front().getIdentifier();
+    for (vector<PeptideIdentification>::iterator it = all_peptide_ids.begin(); it != all_peptide_ids.end(); ++it)
+    {
+      it->setIdentifier(run_identifier);
+      PercolatorFeatureSetHelper::checkExtraFeatures(it->getHits(), extra_features);  // will remove inconsistently available features
+    }
+    
+    // TODO: There should only be 1 ProteinIdentification element in this vector, no need for a for loop
```

Comment:
what about the TODO? is there support for multi run? if yes then remove if not then break with error